File: CS/Q236875/LayoutSettingsPersistingHelper.cs
using System.Collections;
using System;
using System.Collections.Generic;
using System.IO;
using System.Xml.Serialization;
using System.Reflection;
using System.Windows.Forms;
using System.Diagnostics;
using DevExpress.XtraEditors;
using System.ComponentModel;

namespace DXSample {
    public class LayoutSettingsPersistHelper {
        private Dictionary<object, string> fControls = new Dictionary<object, string>();
        IList<ControlLayout> deserializedCache = new List<ControlLayout>();

        public LayoutSettingsPersistHelper(IEnumerable<ControlWithName> controls) {
            foreach (ControlWithName control in controls)
                fControls.Add(control.Control, control.Name);
        }

        public void SaveLayout(string fileName) {
            RefreshCache(fileName);
            foreach (KeyValuePair<object, string> control in fControls)
                SaveLayout(GetControlLayout(control.Key), control.Key);
            if (File.Exists(fileName)) File.Delete(fileName);
            FlushCache(fileName);
            deserializedCache.Clear();
        }

        public void SaveLayout(string fileName, object control) {
            RefreshCache(fileName);
            SaveLayout(GetControlLayout(control), control);
            new FileInfo(fileName).Delete();
            if (File.Exists(fileName)) File.Delete(fileName);
            FlushCache(fileName);
            deserializedCache.Clear();
        }

        public void RestoreLayout(string fileName) {
            RefreshCache(fileName);
            foreach (KeyValuePair<object, string> control in fControls)
                RestoreLayout(GetControlLayout(control.Key), control.Key);
            deserializedCache.Clear();
        }

        public void RestoreLayout(string fileName, object control) {
            RefreshCache(fileName);
            RestoreLayout(GetControlLayout(control), control);
            deserializedCache.Clear();
        }

        private ControlLayout GetControlLayout(object control) {
            ControlLayout layout = null;
            string key = fControls[control];
            foreach (ControlLayout item in deserializedCache)
                if (item.ControlID == key) {
                    layout = item;
                    break;
                }
            if (layout == null) {
                layout = new ControlLayout(key);
                deserializedCache.Add(layout);
            }
            return layout;
        }

        private void SaveLayout(ControlLayout layout, object control) {
            using (MemoryStream stream = new MemoryStream()) {
                try {
                    MethodInfo mi = GetLayoutMethod(control.GetType(), "SaveLayoutToStream");
                    mi.Invoke(control, new object[] { stream });
                    layout.Layout = stream.ToArray();
                } finally { stream.Close(); }
            }
        }

        private void RestoreLayout(ControlLayout layout, object control) {
            if (layout.Layout == null) return;
            using (MemoryStream stream = new MemoryStream(layout.Layout)) {
                try {
                    MethodInfo mi = GetLayoutMethod(control.GetType(), "RestoreLayoutFromStream");
                    mi.Invoke(control, new object[] { stream });
                } finally { stream.Close(); }
            }
        }

        private static MethodInfo GetLayoutMethod(Type type, string methodName) {
            MethodInfo mi = type.GetMethod(methodName,
                BindingFlags.Instance | BindingFlags.Public, null, new Type[] { typeof(Stream) },
                null);
            if (mi == null)
                throw new ArgumentException("The specified control can't save layout");
            return mi;
        }

        private void RefreshCache(string fileName) {
            if (!File.Exists(fileName)) return;
            using (Stream stream = new FileStream(fileName, FileMode.OpenOrCreate)) {
                try {
                    deserializedCache =
                        (IList<ControlLayout>)new XmlSerializer(typeof(List<ControlLayout>)).Deserialize(stream);
                } catch { 
                    deserializedCache.Clear();
                    XtraMessageBox.Show("Operation wasn't completed. The xml file is corrupted");
                } 
                finally { stream.Close(); }
            }
        }

        private void FlushCache(string fileName) {
            using (Stream stream = new FileStream(fileName, FileMode.OpenOrCreate)) {
                try {
                    new XmlSerializer(typeof(List<ControlLayout>)).Serialize(stream, deserializedCache);
                } finally { stream.Close(); }
            }
        }
    }

    public class ControlLayout {
        public ControlLayout(string id) {
            ControlID = id;
        }
        private ControlLayout() { }
        public string ControlID;
        public byte[] Layout;
    }

    public class ControlWithName {
        public ControlWithName(object ctrl, string n) {
            Control = ctrl; Name = n;
        }
        public object Control;
        public string Name;
    }
}

File: VB/Q236875/LayoutSettingsPersistingHelper.vb
Imports Microsoft.VisualBasic
Imports System.Collections
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Xml.Serialization
Imports System.Reflection
Imports System.Windows.Forms
Imports System.Diagnostics
Imports DevExpress.XtraEditors
Imports System.ComponentModel

Namespace DXSample
	Public Class LayoutSettingsPersistHelper
		Private fControls As New Dictionary(Of Object, String)()
		Private deserializedCache As IList(Of ControlLayout) = New List(Of ControlLayout)()

		Public Sub New(ByVal controls As IEnumerable(Of ControlWithName))
			For Each control As ControlWithName In controls
				fControls.Add(control.Control, control.Name)
			Next control
		End Sub

		Public Sub SaveLayout(ByVal fileName As String)
			RefreshCache(fileName)
			For Each control As KeyValuePair(Of Object, String) In fControls
				SaveLayout(GetControlLayout(control.Key), control.Key)
			Next control
			If File.Exists(fileName) Then
				File.Delete(fileName)
			End If
			FlushCache(fileName)
			deserializedCache.Clear()
		End Sub

		Public Sub SaveLayout(ByVal fileName As String, ByVal control As Object)
			RefreshCache(fileName)
			SaveLayout(GetControlLayout(control), control)
			CType(New FileInfo(fileName), FileInfo).Delete()
			If File.Exists(fileName) Then
				File.Delete(fileName)
			End If
			FlushCache(fileName)
			deserializedCache.Clear()
		End Sub

		Public Sub RestoreLayout(ByVal fileName As String)
			RefreshCache(fileName)
			For Each control As KeyValuePair(Of Object, String) In fControls
				RestoreLayout(GetControlLayout(control.Key), control.Key)
			Next control
			deserializedCache.Clear()
		End Sub

		Public Sub RestoreLayout(ByVal fileName As String, ByVal control As Object)
			RefreshCache(fileName)
			RestoreLayout(GetControlLayout(control), control)
			deserializedCache.Clear()
		End Sub

		Private Function GetControlLayout(ByVal control As Object) As ControlLayout
			Dim layout As ControlLayout = Nothing
			Dim key As String = fControls(control)
			For Each item As ControlLayout In deserializedCache
				If item.ControlID = key Then
					layout = item
					Exit For
				End If
			Next item
			If layout Is Nothing Then
				layout = New ControlLayout(key)
				deserializedCache.Add(layout)
			End If
			Return layout
		End Function

		Private Sub SaveLayout(ByVal layout As ControlLayout, ByVal control As Object)
			Using stream As New MemoryStream()
				Try
					Dim mi As MethodInfo = GetLayoutMethod(control.GetType(), "SaveLayoutToStream")
					mi.Invoke(control, New Object() { stream })
					layout.Layout = stream.ToArray()
				Finally
					stream.Close()
				End Try
			End Using
		End Sub

		Private Sub RestoreLayout(ByVal layout As ControlLayout, ByVal control As Object)
			If layout.Layout Is Nothing Then
				Return
			End If
			Using stream As New MemoryStream(layout.Layout)
				Try
					Dim mi As MethodInfo = GetLayoutMethod(control.GetType(), "RestoreLayoutFromStream")
					mi.Invoke(control, New Object() { stream })
				Finally
					stream.Close()
				End Try
			End Using
		End Sub

		Private Shared Function GetLayoutMethod(ByVal type As Type, ByVal methodName As String) As MethodInfo
			Dim mi As MethodInfo = type.GetMethod(methodName, BindingFlags.Instance Or BindingFlags.Public, Nothing, New Type() { GetType(Stream) }, Nothing)
			If mi Is Nothing Then
				Throw New ArgumentException("The specified control can't save layout")
			End If
			Return mi
		End Function

		Private Sub RefreshCache(ByVal fileName As String)
			If (Not File.Exists(fileName)) Then
				Return
			End If
			Using stream As Stream = New FileStream(fileName, FileMode.OpenOrCreate)
				Try
					deserializedCache = CType(New XmlSerializer(GetType(List(Of ControlLayout))).Deserialize(stream), IList(Of ControlLayout))
				Catch
					deserializedCache.Clear()
					XtraMessageBox.Show("Operation wasn't completed. The xml file is corrupted")
				Finally
					stream.Close()
				End Try
			End Using
		End Sub

		Private Sub FlushCache(ByVal fileName As String)
			Using stream As Stream = New FileStream(fileName, FileMode.OpenOrCreate)
				Try
					CType(New XmlSerializer(GetType(List(Of ControlLayout))), XmlSerializer).Serialize(stream, deserializedCache)
				Finally
					stream.Close()
				End Try
			End Using
		End Sub
	End Class

	Public Class ControlLayout
		Public Sub New(ByVal id As String)
			ControlID = id
		End Sub
		Private Sub New()
		End Sub
		Public ControlID As String
		Public Layout() As Byte
	End Class

	Public Class ControlWithName
		Public Sub New(ByVal ctrl As Object, ByVal n As String)
			Control = ctrl
			Name = n
		End Sub
		Public Control As Object
		Public Name As String
	End Class
End Namespace